core/src/actors/session/handlers.rs

```diff
@@ -327,21 +329,55 @@ fn inventory_process_block(_session: &mut Session, _ctx: &mut Context<Session>,
 }
 
 /// Function to process an Inv message
-fn inventory_process_inv(session: &mut Session, _ctx: &mut Context<Session>, inv: &Inv) {
+fn inventory_process_inv(session: &mut Session, ctx: &mut Context<Session>, inv: &Inv) {
     // Check how many of the received inventory vectors need to be requested
     let inv_vectors = &inv.inventory;
 
-    // TODO missing check of how many of these items really need to be requested
-    let missing_inv_vectors = inv_vectors;
+    // Get BlocksManager address
+    let blocks_manager_addr = System::current().registry().get::<BlocksManager>();
 
-    // Check if there are any vectors to be requested
-    if !missing_inv_vectors.is_empty() {
-        // Create GetData message with requested inventory vectors
-        let get_data_msg = WitnetMessage::build_get_data(missing_inv_vectors.to_vec());
+    // Send a message to the BlocksManager to try to add a new block
+    blocks_manager_addr
+        // Send GetConfig message to config manager actor
+        // This returns a Request Future, representing an asynchronous message sending process
+        .send(DiscardExistingInvVectors {
+            inv_vectors: inv_vectors.to_vec(),
+        })
+        // Convert a normal future into an ActorFuture
+        .into_actor(session)
+        // Process the response from the blocks manager
+        // This returns a FutureResult containing the socket address if present
+        .then(|res, _act, _ctx| {
+            // Process the Result<InvVectorsResult, MailboxError>
+            match res {
+                Err(e) => {
+                    error!("Unsuccessful communication with blocks manager: {}", e);
+                    actix::fut::err(())
+                }
+                Ok(res) => match res {
+                    Err(_) => {
+                        error!("Error while filtering inventory vectors");
+                        actix::fut::err(())
+                    }
+                    Ok(res) => actix::fut::ok(res),
+                },
+            }
+        })
+        // Process the received filtered inv elems
+        // This returns a FutureResult containing a success
+        .and_then(|missing_inv_vectors, act, _ctx| {
+            // Check if there are any vectors to be requested
+            if !missing_inv_vectors.is_empty() {
+                // Create GetData message with requested inventory vectors
+                let get_data_msg = WitnetMessage::build_get_data(missing_inv_vectors.to_vec());
```

Comment:
Question: what would happen if we built a GetData message with an empty vector? Will it crash?

Reply:
It shouldn't crash (we're testing it though :wink: ). Anyway, I think it may be unnecessary to send a `GetData` protocol message through the network if there's nothing to request

Reply:
An empty `GetData` is sent through the network:
```
-----> Session (V4(127.0.0.1:60208)) sending message: Message { kind: GetData(GetData { inventory: [] }), magic: 43981 }
```
So should we somehow move the check inside the message builder, so that it is impossible to send an empty `GetData` message. The same logic may also apply to `Inv` messages.